values.yaml

```diff
@@ -1008,3 +1030,82 @@ csi:
   # See https://www.vaultproject.io/docs/platform/k8s/csi/configurations#command-line-arguments
   # for the available command line flags.
   extraArgs: []
+
+telemetry:
+  # Vault is able to collect and publish various runtime metrics.
+  # Enabling this feature requires setting adding `telemetry{}` stanza to
+  # the Vault configuration. There are a few examples included in the `config` sections above.
+  #
+  # For more information see:
+  # https://www.vaultproject.io/docs/configuration/telemetry
+  # https://www.vaultproject.io/docs/internals/telemetry
+
+  # Enable support for the Prometheus Operator. Currently, this chart does not support
+  # authenticating to Vault's metrics endpoint, so the following `telemetry{}` must be included
+  # in the `listener "tcp"{}` stanza
+  #  telemetry {
+  #    unauthenticated_metrics_access = "true"
+  #  }
+  #
+  # See the `standalone.config` for a more complete example of this.
+  #
+  # In addition, a top level `telemetry{}` stanza must also be included in the Vault configuration:
+  #
+  # example:
+  #  telemetry {
+  #    prometheus_retention_time = "30s",
+  #    disable_hostname = true
+  #  }
+  #
+  # The Prometheus operator *must* be installed before enabling this feature,
+  # if not the chart will fail to install due to missing CustomResourceDefinitions
+  # provided by the operator.
+  #
+  # Instructions on how to install the Helm chart can be found here:
+  #  https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack
+  # More information can be found here:
+  #  https://github.com/prometheus-operator/prometheus-operator
+  #  https://github.com/prometheus-operator/kube-prometheus
+  prometheusOperator:
+    # Enable to deploy Prometheus Operator CustomResources
+    enabled: false
+
+    serviceMonitor:
+      # Selector labels to add to the ServiceMonitor.
+      # When empty, defaults to:
+      #  release: prometheus
+      selectors: {}
+
+      # Interval at which Prometheus scrapes metrics
+      interval: 30s
+
+      # Timeout at which Prometheus timeouts scrape run
+      scrapeTimeout: 10s
+
+    prometheusRules:
+      # Deploy the PrometheusRule custom resource for AlertManager based alerts.
+      # Requires that AlertManager is properly deployed.
+      enabled: false
+
+      # Selector labels to add to the ServiceMonitor.
+      # When empty, defaults to:
+      #  release: prometheus
+      #  app: kube-prometheus-stack
+      selectors: {}
+
+      # Some example rules.
+      rules: {}
```

Comment:
I wonder if this should somehow be further divided into a server telemetry section? Just thinking that what's being templated here is specific to Vault server, and if we ever wanted to add a serviceMonitor for the vault-k8s injector for example, we'd need another set of these options.

Reply:
Good call. What do you think of the following structure?
```
  prometheusOperator:
    serviceMonitor:
      server:
        # Enable deployment of the Vault Server ServiceMonitor CustomResource.
        enabled: false

        # Selector labels to add to the ServiceMonitor.
        # When empty, defaults to:
        #  release: prometheus
        selectors: {}

        # Interval at which Prometheus scrapes metrics
        interval: 30s

        # Timeout for Prometheus scrapes
        scrapeTimeout: 10s

    prometheusRules:
      server:
        # Deploy the PrometheusRule custom resource for AlertManager based alerts.
        # Requires that AlertManager is properly deployed.
        enabled: false

[...]
```

We could then extend it to support other `ServiceMonitor` deployments etc.

Reply:
Implemented in d6b9c532b7423921f002fc22a1a6b5ab02071268

Reply:
Yeah that structure should work. We could probably remove either `telemetry` or `prometheusOperator` as well to trim the hierarchy a bit.

Reply:
I like the idea of dropping the latter.

Reply:
Fixed in 4886b1469102e0ce906f2f0ce015872f00b1676c